app/src/main/java/com/pileproject/drive/execution/ExecutionThread.java

```diff
@@ -25,119 +25,78 @@
-
-import java.util.ArrayList;
-import java.util.Map;
-
-import trikita.log.Log;
 
 /**
- * Sub thread to execute program
- *
- * @author <a href="mailto:tatsuyaw0c@gmail.com">Tatsuya Iwanari</a>
- * @version 1.0 7-July-2013
+ * A Thread class to execute program.
  */
 public class ExecutionThread extends Thread {
-    public static final int START_THREAD = 1000;
-    public static final int EMPHASIZE_BLOCK = 1001;
-    public static final int END_THREAD = 1002;
-    public static final int CONNECTION_ERROR = 1003;
-    private static int mThreadNum = 0;
-    private boolean mHalt = false;
+    public static final int START_THREAD        = 1000;
+    public static final int EMPHASIZE_BLOCK     = 1001;
+    public static final int END_THREAD          = 1002;
+    public static final int CONNECTION_ERROR    = 1003;
     private ProgramDataManager mManager;
     private ExecutionCondition mCondition;
     private MachineController mController;
     private Handler mUiHandler;
-    private boolean mStop;
+    private boolean mStop = false;
+    private boolean mHalt = false;
 
     /**
      * Constructor
      *
-     * @param context   The context of Activity that calls this thread
-     * @param uiHandler Handler for controlling the showing progress layout
+     * @param uiHandler a handler for communicating the execution activity
+     * @param controller a controller to control machines (e.g., NXT/EV3)
      */
-    public ExecutionThread(Context context, Handler uiHandler, MachineController controller) {
+    public ExecutionThread(Handler uiHandler, MachineController controller) {
         super();
         mManager = ProgramDataManager.getInstance();
-        mCondition = new ExecutionCondition();
         mController = controller;
         mUiHandler = uiHandler;
     }
 
-    private boolean isExecutable() {
-        synchronized (this) {
-            return (mThreadNum == 0);
-        }
-    }
-
     @Override
     public void run() {
         sendState(START_THREAD);
 
-        // Only one thread can be executed
-        synchronized (this) {
-            if (isExecutable()) {
-                mThreadNum = 1;
-            } else {
-                return;
-            }
-        }
-
-        mCondition.blocks = mManager.loadExecutionProgram();
-        boolean isStopped = false;
+        mCondition = new ExecutionCondition(mManager.loadExecutionProgram());
+        boolean hasStopped = false;
         try {
-            for (mCondition.programCount = 0;
-                 mCondition.programCount < mCondition.blocks.size();
-                 mCondition.programCount++) {
-                // Halt execution
+            while (!mCondition.hasProgramFinished()) {
+
+                // halt execution
                 if (mHalt) {
                     break;
                 }
 
-                // Stop temporarily
+                // stop temporarily
                 if (mStop) {
```

Comment:
It's really painful to understand this if block.
If <code>mController.halt()</code> is idempotent, you could simplify this?